server/src/infra/ldap_handler.rs

```diff
@@ -211,6 +224,216 @@ fn root_dse_response(base_dn: &str) -> LdapOp {
                 atype: "isGlobalCatalogReady".to_string(),
                 vals: vec![b"false".to_vec()],
             },
+            LdapPartialAttribute {
+                atype: "subschemaSubentry".to_string(),
+                vals: vec![b"cn=Subschema".to_vec()],
+            },
+        ],
+    })
+}
+
+pub struct ObjectClassList(Vec<LdapObjectClass>);
+
+impl From<Vec<LdapObjectClass>> for ObjectClassList {
+    fn from(object_class: Vec<LdapObjectClass>) -> Self {
+        ObjectClassList(object_class)
+    }
+}
+
+// See RFC4512 section 4.2.1 "objectClasses"
+impl ObjectClassList {
+    fn format_for_ldap_schema_description(&self) -> String {
+        self.0
+            .iter()
+            .map(|c| format!("'{}'", c))
+            .unique()
+            .collect::<Vec<_>>()
+            .join(" ")
+    }
+}
+
+// See RFC4512 section 4.2 "Subschema Subentries"
+// This struct holds all information on what attributes and objectclasses are present on the server.
+// It can be used to 'index' a server using a LDAP subschema call.
+pub struct LdapSchemaDescription {
+    base: Schema,
+    user_object_classes: ObjectClassList,
+    group_object_classes: ObjectClassList,
+}
+
+impl Deref for LdapSchemaDescription {
+    type Target = Schema;
+
+    fn deref(&self) -> &Self::Target {
+        &self.base
+    }
+}
+
+impl LdapSchemaDescription {
+    fn from(schema: &Schema) -> Self {
```

Comment:
This should take the schema by value, to let the caller decide whether to clone it or not, since we need a value anyway.
Even better, this can take a `PublicSchema` by value, so we never forget to do the conversion. Semantically, it operates on the public schema anyway.